polkadot/node/core/pvf/execute-worker/src/lib.rs

```diff
@@ -149,10 +154,26 @@ pub fn worker_entrypoint(
 			let worker_pid = process::id();
 			let artifact_path = worker_dir::execute_artifact(&worker_dir_path);
 
-			let Handshake { executor_params } = recv_execute_handshake(&mut stream)?;
+			let Handshake { executor_params } = match recv_execute_handshake(&mut stream) {
```

Comment:
Is this change (and the following similar changes) worth it? It's a complication, not a big one, but still, a couple of dozens of lines. Does it make things better? I believe not. Neither handshake nor request contains untrusted data, and an error would mean everything went terribly wrong. It doesn't make much sense to me to communicate that back to the host. If the communication is so much broken at this point, is it even possible to communicate it back?

Reply:
Good point! I can revert the `HostCommunication` errors here.

Reply:
Looks like we can use `map_err` to save one line. :P

Reply:
Okay, one more thought, then. The worker main loop is a closure that returns `io::Result<Never>` to the `run_worker()` function. Maybe those errors should be handled on the `run_worker()` level? That would still allow us to use `?` in the main loop and keep error handling simple. Not sure it makes total sense, but it probably is worth considering.

Reply:
Thanks, that's a great idea. Right now we quit the workers on `io::Error` but continue looping (most of the time) on other kinds of errors. But it should be that if any error occurs, the worker always quits. Then we could make `run_worker` generic over both ExecuteResult and PrepareResult.

The question is, can the worker always just quit if an error occurred?
- execute: yes, an `Err` response always means that the worker dies (at least, it will after this refactor)
- prepare: no, it’s not so clear in the un-refactored code here but we return an error if compilation fails. We still want to keep the current behavior (keeping the worker alive in that case).

I’d like to refactor the prepare errors which would enable the `run_worker` refactor. But it’s more work which is probably going to become obsolete soon. It would also create more merge conflicts with the other PRs. So I will drop this idea to refactor `run_worker` as well, although it’s a good idea and I would like to do it. 

We can raise a follow-up just in case, and I'll push what I already implemented. There is some duplication of code, but this code may not need to be maintained for much longer, anyway, and it should make things a bit more robust in the meantime.

Reply:
I realize that `send_error` wrapping the error in an `io::Error` is pretty weird. I thought about doing the refactor described above, but then I remembered that killing the prepare-worker on error causes the host to [send a `FromPool::Rip` signal](https://github.com/paritytech/polkadot-sdk/blob/dbd8d20b25b29abf138e3ad35c40df4c55bd7c5c/polkadot/node/core/pvf/src/prepare/pool.rs#L201-L205). This may race with the actual error sent to the host from the worker before it dies, though I'm not sure what exactly happens. (Note that there is no race in the equivalent [execute-worker code](https://github.com/paritytech/polkadot-sdk/blob/3d847e333d4da3c3e234b37723db9b99948c3a6f/polkadot/node/core/pvf/src/execute/queue.rs#L275-L279), so the code we have now works - internal errors should get reported.) Clearly this also requires a deeper look, so it should also be a follow-up. It can be addressed later if some of this code remains when PolkaVM is integrated.